### ISSUE TITLE
Support for lua_yield

### ISSUE THREAD

**Issue by firas-assaad** [ThePhD/sol2#589]
It would be nice to have the ability to yield from a C++ function. This is supported by the Lua API through the `lua_yield`/`lua_yieldk` functions, and can also be found in some other libraries such as Luabind or luacppinterface. According to #272 this method is restrictive and unsafe.

My particular use case is a scheduling system similar to the one described in [this article](http://lua.space/gamedev/using-lua-coroutines-to-create-rpg), but with the ability to have Lua code wait on operations defined in C++. Here is the [Luabind-based implementation](https://github.com/firas-assaad/xd/blob/master/src/lua/scheduler.cpp#L96) I currently use. This allows calling a function such as `show_text(...)` in Lua, where efficient text display and waiting for user input are implemented in C++, and then having other Lua scripts continue to run cooperatively. 

I am sure there are many other use cases. For example, having this feature allows writing asynchronous Lua code that depends on a slow algorithm or IO operation, such as scheduling network connections or getting data from a database. Then you'd be able to write:

    local task1 = slow_cpp_operation1()
    local task2 = slow_cpp_operation2()
    ...
    -- some more code
    ...
    local result = task1:wait_for_value() + task2:wait_for_value()

As for the implementation, Luabind enables this with the [yield policy](http://luabind.sourceforge.net/docs.html#policies_yield) when mapping a function, and luacppinterface has a [CreateYieldingFunction](https://github.com/davidsiaw/luacppinterface#luacoroutine-object) method.

Not sure if it'd be possible at all, but currently this is the only thing holding me back from switching to this awesome library.

**Comment by ThePhD** [comment]
It almost looks like these implementations expect your C++ function to finish before it calls yield, rather just allowing you to yield at any point in the function (which is what I was trying to implement the last time and could not because that essentially required "saving the execution stack", which required third-party libs like Boost.Fiber).

If we make the stipulation that the bound routine must finish, then this is implementable. In a way.

Is this what is desired?

**Comment by firas-assaad** [comment]
Yes you are right about the other implementations. Yielding when the function finishes would be sufficient for the use cases I mentioned (C++ function registers a callable predicate with the scheduler before yielding on exit. The predicate is checked every frame to resume the thread that yielded).

**Comment by firas-assaad** [comment]
Yes you are right about the other implementations. Yielding when the function finishes would be sufficient for the use cases I mentioned (C++ function registers a callable predicate with the scheduler before yielding on exit. The predicate is checked every frame to resume the thread that yielded).

**Comment by ThePhD** [comment]
This might be simple to implement, but I won't get to it until later.

**Comment by ThePhD** [comment]
Preliminary implementation seems okay:

![Wee](https://i.imgur.com/LgOWNoP.png)

**Comment by ThePhD** [comment]
@firas-assaad Try your code now. Wrap the function you want to yield from with `sol::yielding( func... )`. Let me know what doesn't compile / breaks.

**Comment by firas-assaad** [comment]
That's great, thanks a lot! It will take quite some time to rewrite my current implementation, but I tried it with a simpler version and it works perfectly (even with concurrent Lua threads): [gist](https://gist.github.com/firas-assaad/efed5564b3ba5829b99e72df82196ea6)

Code like this:
    print('Lua - Before first wait')
    wait(3)
    print('Lua - After wait, before show_text')
    show_text('hello')
    print('Lua - After show_text')

Outputs the following:
    Lua - Before wait
    C++ - Waiting for 3 seconds
    Lua - After wait, before show_text
    C++ - Showing text: hello
    Lua - After show_text

I will let you know if I run into any other issues.

**Comment by ThePhD** [comment]
Alrighty. Feel free to ping this issue if you come across anything strange.

**Comment by davidchisnall** [comment]
The example from the gist behaves very strangely if user code creates its own coroutines.  My mental model for this is:

 - User code should be able to create coroutines.
 - User code can resume other coroutines that have yielded, with no intervention of a scheduler.
 - If you want to implement a scheduler then your blocking functions explicitly yield to the scheduler.
 - When the scheduler runs, it resumes one of the coroutines that has blocked.

To make this work, I'd expect `sol::yielding` would give me a mechanism to capture the calling coroutine / thread that I need to resume.  Something along the lines of:

```
// To implement the blocking function:
lua["blockingFunction"] = sol::yielding([&](sol::coroutine resume) { capture(resume); };
```

The scheduler would then just call the captured copy of `resume` to provide the result of the function.  If at all possible, I don't want to explicitly create coroutines / threads in the C++ side at all, I want this to be something that Lua programmers can do if they want to call blocking functions.

I've tried modifying the example to replace `coroutine.create` with something that creates a new `sol::thread`, transfers ownership of the function to the new thread (note: I didn't find anything about how to do this in the docs, it was trial and error based on one of the examples), and then returning a function that, once called, marks the new function as runnable and yields.  This works, but it feels like I'm fighting Sol at every step of the way and I feel that there must be a better way.
